### PR TITLE
kv/MemDB: add perfcounter.

### PR DESCRIPTION
**Pull request by majianpeng**
Signed-off-by: Jianpeng Ma jianpeng.ma@intel.com

### CODE REVIEW

**Comment by yuyuyu101** [comment]
rebase

**Comment by liewegas** [comment]
please rebase?

**Comment by liewegas** [comment]
@majianpeng ping

**Comment by majianpeng** [comment]
update

**Comment by liewegas** [comment]
fails ceph_test_keyvaluedb
/a/sage-2017-05-03_20:20:09-rados-wip-sage-testing---basic-smithi/1098315

**Comment by majianpeng** [comment]
In fact, i don't change anything. I only rebase and make it rebuild.

**Comment by liewegas** [comment]
ping.. needs rebase?

**Comment by majianpeng** [comment]
update!

**Comment by liewegas** [comment]
needs rebase

**Comment by majianpeng** [comment]
update. please review. Thanks!

**Comment by liewegas** [comment]
```
2018-06-26T17:00:52.432 INFO:teuthology.orchestra.run.smithi162.stderr: ceph version 14.0.0-834-gf206106 (f206106a05c4e7b9136f6e05d7a71e8d398b7d35) nautilus (dev)
2018-06-26T17:00:52.432 INFO:teuthology.orchestra.run.smithi162.stderr: 1: (()+0x11390) [0x7f48454fc390]
2018-06-26T17:00:52.432 INFO:teuthology.orchestra.run.smithi162.stderr: 2: (PerfCountersCollection::remove(PerfCounters*)+0x3e) [0x7f483bf061de]
2018-06-26T17:00:52.433 INFO:teuthology.orchestra.run.smithi162.stderr: 3: (MemDB::~MemDB()+0x1e) [0x564921d8616e]
2018-06-26T17:00:52.433 INFO:teuthology.orchestra.run.smithi162.stderr: 4: (MemDB::~MemDB()+0x9) [0x564921d863f9]
2018-06-26T17:00:52.433 INFO:teuthology.orchestra.run.smithi162.stderr: 5: (KVTest_OpenClose_Test::TestBody()+0xca) [0x564921d697aa]
2018-06-26T17:00:52.433 INFO:teuthology.orchestra.run.smithi162.stderr: 6: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x4a) [0x564921dcc93a]
2018-06-26T17:00:52.434 INFO:teuthology.orchestra.run.smithi162.stderr: 7: (testing::Test::Run()+0xba) [0x564921dc09ca]
2018-06-26T17:00:52.434 INFO:teuthology.orchestra.run.smithi162.stderr: 8: (testing::TestInfo::Run()+0xaf) [0x564921dc0aaf]
2018-06-26T17:00:52.434 INFO:teuthology.orchestra.run.smithi162.stderr: 9: (testing::TestCase::Run()+0xb5) [0x564921dc0be5]
2018-06-26T17:00:52.434 INFO:teuthology.orchestra.run.smithi162.stderr: 10: (testing::internal::UnitTestImpl::RunAllTests()+0x468) [0x564921dc10d8]
2018-06-26T17:00:52.435 INFO:teuthology.orchestra.run.smithi162.stderr: 11: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x4a) [0x564921dccdfa]
2018-06-26T17:00:52.435 INFO:teuthology.orchestra.run.smithi162.stderr: 12: (testing::UnitTest::Run()+0x5d) [0x564921dc125d]
2018-06-26T17:00:52.435 INFO:teuthology.orchestra.run.smithi162.stderr: 13: (main()+0xe9) [0x564921ca5ec9]
```
/a/sage-2018-06-26_15:58:10-rados-wip-sage3-testing-2018-06-26-0616-distro-basic-smithi/2706741

**Comment by majianpeng** [comment]
@liewegas . Sorry for later response. I think before remove perfcounter, it should judge perfcounter != NULL. Because in func __init maybe early return.

**Comment by liewegas** [comment]
```
2018-08-08T04:12:16.569 INFO:teuthology.orchestra.run.smithi031.stderr:    -5> 2018-08-08 04:12:16.356 7f3159cf2ec0  4 rocksdb: [/build/ceph-14.0.0-2033-ge586fe3/src/rocksdb/db/db_impl.cc:252] Shutdown: canceling all background work
2018-08-08T04:12:16.569 INFO:teuthology.orchestra.run.smithi031.stderr:    -4> 2018-08-08 04:12:16.356 7f3159cf2ec0  4 rocksdb: [/build/ceph-14.0.0-2033-ge586fe3/src/rocksdb/db/db_impl.cc:398] Shutdown complete
2018-08-08T04:12:16.569 INFO:teuthology.orchestra.run.smithi031.stderr:    -3> 2018-08-08 04:12:16.360 7f3159cf2ec0 -1 WARNING: experimental feature 'memdb' is enabled
2018-08-08T04:12:16.569 INFO:teuthology.orchestra.run.smithi031.stderr:Please be aware that this feature is experimental, untested,
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:unsupported, and may result in data corruption, data loss,
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:and/or irreparable damage to your cluster.  Do not use
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:feature with important data.
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:    -2> 2018-08-08 04:12:16.360 7f3159cf2ec0  1 memdb: _init
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:    -1> 2018-08-08 04:12:16.364 7f3159cf2ec0 -1 *** Caught signal (Segmentation fault) **
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr: in thread 7f3159cf2ec0 thread_name:ceph_test_keyva
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr:
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr: ceph version 14.0.0-2033-ge586fe3 (e586fe3e94d88f867f9f32ea2bbea56674448dcb) nautilus (dev)
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr: 1: (()+0x12890) [0x7f31598df890]
2018-08-08T04:12:16.570 INFO:teuthology.orchestra.run.smithi031.stderr: 2: (PerfCounters::inc(int, unsigned long)+0) [0x7f31507bfaa0]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 3: (MemDB::submit_transaction(std::shared_ptr<KeyValueDB::TransactionImpl>)+0x2b1) [0x5557c5f5e9f1]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 4: (MemDB::submit_transaction_sync(std::shared_ptr<KeyValueDB::TransactionImpl>)+0x70) [0x5557c5f576c0]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 5: (KVTest_OpenWriteRead_Test::TestBody()+0x2de) [0x5557c5f4522e]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 6: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x4a) [0x5557c5fa019a]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 7: (testing::Test::Run()+0xba) [0x5557c5f9422a]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 8: (testing::TestInfo::Run()+0xaf) [0x5557c5f9430f]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 9: (testing::TestCase::Run()+0xb5) [0x5557c5f94445]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 10: (testing::internal::UnitTestImpl::RunAllTests()+0x468) [0x5557c5f94938]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 11: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x4a) [0x5557c5fa065a]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 12: (testing::UnitTest::Run()+0x5d) [0x5557c5f94abd]
2018-08-08T04:12:16.571 INFO:teuthology.orchestra.run.smithi031.stderr: 13: (main()+0x426) [0x5557c5e7a516]
2018-08-08T04:12:16.572 INFO:teuthology.orchestra.run.smithi031.stderr: 14: (__libc_start_main()+0xe7) [0x7f314f2b4b97]
2018-08-08T04:12:16.572 INFO:teuthology.orchestra.run.smithi031.stderr: 15: (_start()+0x2a) [0x5557c5f3d82a]
2018-08-08T04:12:16.572 INFO:teuthology.orchestra.run.smithi031.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2018-08-08T04:12:16.572 INFO:teuthology.orchestra.run.smithi031.stderr:
```
from ceph_test_keyvaluedb (you can run this locally)

File: src/kv/MemDB.cc [hunk 1]
```diff
@@ -139,15 +139,23 @@ int MemDB::_init(bool create)
     if (r < 0) {
       r = -errno;
       if (r != -EEXIST) {
-        derr << __func__ << " mkdir failed: " << cpp_strerror(r) << dendl;
-        return r;
+	derr << __func__ << " mkdir failed: " << cpp_strerror(r) << dendl;
+	return r;
       }
-      return 0; // ignore EEXIST
+      r = 0; // ignore EEXIST
```
Comment (majianpeng) [inline src/kv/MemDB.cc]: @liewegas . The bug cause by this.  If r==EEXIST, it return and do continue. for this case, logger isn't init.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-08-17_15:09:18-rados-wip-kefu-testing-2018-08-17-2127-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-08-18_16:46:03-rados-wip-kefu-testing-2018-08-18-2315-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-08-19_13:56:54-rados-wip-kefu-testing-2018-08-19-1812-distro-basic-mira/

the failure is tracked by http://tracker.ceph.com/issues/24485

**Pull request merged**